File: explainer.js
/**
 * AI Resume Screening Tool - Explainer Panel Module
 */

const explainer = {
    init: () => {
        const closeBtn = document.getElementById('drawer-close');
        const overlay = document.getElementById('drawer-overlay');
        
        if (!closeBtn) return;
        
        closeBtn.addEventListener('click', explainer.hide);
        overlay.addEventListener('click', explainer.hide);
    },

    show: (candidateName) => {
        const res = app.state.results.find(r => r.candidate_name === candidateName);
        if (!res) return;

        const drawer = document.getElementById('explainer-drawer');
        const overlay = document.getElementById('drawer-overlay');
        const content = document.getElementById('drawer-content');

        content.innerHTML = `
            <!-- Total Score & Gauges -->
            <div class="summary-box">
                <strong>AI Summary:</strong> ${res.summary}
            </div>

            <div class="score-gauges">
                <div class="gauge-card">
                    <div class="gauge-value">${res.score_breakdown.keyword.raw.toFixed(0)}</div>
                    <div class="gauge-label">Skills Match</div>
                    <div class="gauge-sub">${res.matched_keywords.length} found</div>
                </div>
                <div class="gauge-card">
                    <div class="gauge-value">${res.score_breakdown.experience.raw.toFixed(0)}</div>
                    <div class="gauge-label">Experience</div>
                    <div class="gauge-sub">${res.years_experience} Years</div>
                </div>
                <div class="gauge-card">
                    <div class="gauge-value">${res.score_breakdown.education.raw.toFixed(0)}</div>
                    <div class="gauge-label">Education</div>
                    <div class="gauge-sub">${res.education_level}</div>
                </div>
            </div>

            <!-- Strengths & Gaps -->
            <div>
                <div class="drawer-section-title">Key Strengths</div>
                <div class="sg-list">
                    ${res.strengths.map(s => `
                        <div class="sg-item strength">
                            <i class="fas fa-plus-circle sg-dot"></i>
                            <span>${s}</span>
                        </div>
                    `).join('')}
                </div>
            </div>

            <div>
                <div class="drawer-section-title">Missing Criteria</div>
                <div class="sg-list">
                    ${res.gaps.length > 0 ? res.gaps.map(g => `
                        <div class="sg-item gap">
                            <i class="fas fa-minus-circle sg-dot"></i>
                            <span>${g}</span>
                        </div>
                    `).join('') : '<div style="font-size:12px; color:var(--text-muted)">No major gaps identified.</div>'}
                </div>
            </div>

            <!-- Keywords chips -->
            <div>
                <div class="drawer-section-title">Skills Breakdown</div>
                <div class="kw-chips">
                    ${res.matched_keywords.map(k => `
                        <div class="kw-chip matched">
                            <i class="fas fa-check chip-icon"></i>
                            ${k.keyword}
                        </div>
                    `).join('')}
                    ${res.missing_keywords.map(k => `
                        <div class="kw-chip missing">
                            <i class="fas fa-times chip-icon"></i>
                            ${k.keyword}
                        </div>
                    `).join('')}
                </div>
            </div>

            <!-- Context Snippets -->
            <div>
                <div class="drawer-section-title">Verified Snippets from Resume</div>
                <div class="snippets-container">
                    ${Object.entries(res.snippets).length > 0 ? 
                        Object.entries(res.snippets).map(([kw, snips]) => `
                            ${snips.map(s => `
                                <div class="snippet-item">
                                    <div class="snippet-kw">${s.keyword}</div>
                                    <div class="snippet-text">"...${explainer.highlight(s.text, s.keyword)}..."</div>
                                </div>
                            `).join('')}
                        `).join('') : 
                        '<div style="font-size:12px; color:var(--text-muted)">No snippets extracted for this candidate.</div>'}
                </div>
            </div>
        `;

        drawer.classList.add('open');
        overlay.classList.add('open');
    },

    highlight: (text, kw) => {
        const regex = new RegExp(`(${kw})`, 'gi');
        return text.replace(regex, '<mark>$1</mark>');
    },

    hide: () => {
        document.getElementById('explainer-drawer').classList.remove('open');
        document.getElementById('drawer-overlay').classList.remove('open');
    }
};

document.addEventListener('DOMContentLoaded', explainer.init);

File: explainer.py
import re

_SNIPPET_WINDOW = 80  # chars around match to show as context

def get_keyword_snippets(text, keywords):
    """
    For each matched keyword, extract up to 2 snippet contexts from the resume.
    Returns dict: {keyword_lower: [snippet1, snippet2, ...]}
    """
    snippets = {}
    for kw_obj in keywords:
        kw = kw_obj.get('keyword', '').strip().lower()
        if not kw:
            continue
        pattern = re.compile(re.escape(kw), re.IGNORECASE)
        matches = list(pattern.finditer(text))
        kw_snippets = []
        for m in matches[:3]:  # max 3 snippets per keyword
            start = max(0, m.start() - _SNIPPET_WINDOW)
            end = min(len(text), m.end() + _SNIPPET_WINDOW)
            snippet = text[start:end].replace('\n', ' ').strip()
            # Trim to sentence boundaries if possible
            snippet = re.sub(r'\s+', ' ', snippet)
            kw_snippets.append({
                'text': snippet,
                'keyword': kw_obj.get('keyword', kw),
                'start_offset': m.start(),
            })
        if kw_snippets:
            snippets[kw] = kw_snippets
    return snippets

def build_explanation(text, score_result):
    """
    Build a full explainability report for a single resume.
    Takes scoring result dict and adds snippet context.
    """
    matched = score_result.get('matched_keywords', [])
    missing = score_result.get('missing_keywords', [])
    breakdown = score_result.get('score_breakdown', {})

    # Get snippets for matched keywords
    snippets = get_keyword_snippets(text, matched)

    explanation = {
        'matched_keywords': matched,
        'missing_keywords': missing,
        'snippets': snippets,
        'score_breakdown': breakdown,
        'summary': _build_summary(score_result),
        'strengths': _build_strengths(score_result),
        'gaps': _build_gaps(score_result),
    }
    return explanation

def _build_summary(score_result):
    total = score_result.get('total_score', 0)
    matched = score_result.get('matched_keywords', [])
    missing = score_result.get('missing_keywords', [])
    years = score_result.get('years_experience', 0)
    edu = score_result.get('education_level', 'Not Detected')

    if total >= 80:
        strength = 'Excellent'
    elif total >= 60:
        strength = 'Good'
    elif total >= 40:
        strength = 'Average'
    else:
        strength = 'Below Average'

    return (
        f"{strength} match ({total:.1f}/100). "
        f"Matched {len(matched)} of {len(matched)+len(missing)} keywords. "
        f"{years} year(s) of experience detected. "
        f"Education: {edu}."
    )

def _build_strengths(score_result):
    strengths = []
    matched = score_result.get('matched_keywords', [])
    years = score_result.get('years_experience', 0)
    edu = score_result.get('education_level', 'Not Detected')

    high_value = [k for k in matched if k.get('weight', 1.0) >= 1.5]
    if high_value:
        kw_names = ', '.join(k['keyword'] for k in high_value[:5])
        strengths.append(f"Has high-priority skills: {kw_names}")

    if years >= 5:
        strengths.append(f"Strong experience: {years} years")
    elif years >= 2:
        strengths.append(f"Solid experience: {years} years")

    if edu.lower() in ('phd', 'ph.d', 'doctorate', 'doctor of philosophy'):
        strengths.append("Doctoral-level education")
    elif edu.lower() in ('master', 'msc', 'mba', 'm.sc', 'm.s.'):
        strengths.append("Master's degree education")
    elif edu.lower() in ('bachelor', 'b.sc', 'b.s.', 'b.e.', 'b.tech'):
        strengths.append("Bachelor's degree education")

    return strengths

def _build_gaps(score_result):
    gaps = []
    missing = score_result.get('missing_keywords', [])
    years = score_result.get('years_experience', 0)

    critical_missing = [k for k in missing if k.get('weight', 1.0) >= 1.5]
    if critical_missing:
        kw_names = ', '.join(k['keyword'] for k in critical_missing[:5])
        gaps.append(f"Missing high-priority skills: {kw_names}")

    other_missing = [k for k in missing if k.get('weight', 1.0) < 1.5]
    if other_missing:
        kw_names = ', '.join(k['keyword'] for k in other_missing[:5])
        gaps.append(f"Missing skills: {kw_names}")

    if years < 2:
        gaps.append(f"Limited experience: only {years} year(s) detected")

    return gaps
